### PR TITLE
feat: convolutional vae for multivariate time series

### PR DESCRIPTION
**Pull request by ab93**
- Introduce variational autoencoder for multivariate time series data
- Customized trainer for vae
- Causal convolutional modules
- Improved training console logging

Exploration in https://github.com/numaproj/numalogic-benchmarks/pull/3

### CODE REVIEW

File: numalogic/models/vae/base.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    if loss_fn == "huber":
+        return F.huber_loss
+    if loss_fn == "l1":
+        return F.l1_loss
+    if loss_fn == "mse":
+        return F.mse_loss
+    raise ValueError(f"Unsupported loss function provided: {loss_fn}")
+
+
+class BaseVAE(TorchModel):
```
Comment (s0nicboOm) [inline numalogic/models/vae/base.py:19]: can't we use `BaseAE` class here too?

Comment (s0nicboOm) [reply numalogic/models/vae/base.py:19]: Loss function, optimizers all seem to be same.

**Review by s0nicboOm — APPROVED** [review]
minor comments. 
LGTM

File: numalogic/models/vae/base.py (hunk 1 above)
Comment (ab93) [reply numalogic/models/vae/base.py:19]: It will be different moving on is my understanding. Even now, the validation and prediction steps are different

File: numalogic/models/vae/variants/conv.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+        n_features: num of features
+        num_conv_filters: number of convolutional filters
+        latent_dim: latent dimension
+    """
+
+    def __init__(self, seq_len: int, n_features: int, num_conv_filters: int, latent_dim: int):
+        super().__init__()
+        self.seq_len = seq_len
+        self.n_features = n_features
+        self.fc = nn.Linear(latent_dim, num_conv_filters * 6)
```
[the remark this answers is not in the document]
Comment (ab93) [reply numalogic/models/vae/variants/conv.py:118]: Gave the best empirical results. Basically here we need to unflatten the features. Moving on, I think we should use the number of features in the dataset to determine that. However, there are no wrong answers, here.
